### PR TITLE
Fix mistake with loading an order in Invoice model

### PR DESCRIPTION
**Pull request by TomashKhamlai**
Should allow you to get an extension attributes from Order using the Invoice model.

### Description
In this pull request was changed the way of getting information about the Invoice.

### Fixed Issues
1. magento/magento2#14424: Mistake with loading an order in Invoice model

### Manual testing scenarios
1. Check that now you can get extension attributes using WebAPI client
2. Check that Magento Swagger also returns extension attributes without error.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=14450) <br/>All committers have signed the CLA.

File: app/code/Magento/Sales/Model/Order/Invoice.php [hunk 1]
```diff
@@ -236,7 +236,7 @@ public function setOrder(\Magento\Sales\Model\Order $order)
     public function getOrder()
     {
         if (!$this->_order instanceof \Magento\Sales\Model\Order) {
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Model/Order/Invoice.php:238]: Since `OrderRepository` may return any class implementing `OrderInterface` this check may be false-negative in some scenarios on customized system.

File: app/code/Magento/Sales/Model/Order/Invoice.php [hunk 2]
```diff
@@ -142,7 +142,7 @@ public function __construct(
         \Magento\Framework\Api\ExtensionAttributesFactory $extensionFactory,
         AttributeValueFactory $customAttributeFactory,
         \Magento\Sales\Model\Order\Invoice\Config $invoiceConfig,
-        \Magento\Sales\Model\OrderFactory $orderFactory,
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Model/Order/Invoice.php:145]: It's not allowed to change constructor signature, especially for classes marked with `@api` annotation.
More information is available here: http://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/index.html

File: app/code/Magento/Sales/Model/Order/Invoice.php [hunk 3]
```diff
@@ -236,7 +236,7 @@ public function setOrder(\Magento\Sales\Model\Order $order)
     public function getOrder()
     {
         if (!$this->_order instanceof \Magento\Sales\Model\Order) {
-            $this->_order = $this->_orderFactory->create()->load($this->getOrderId());
+            $this->_order = $this->orderRepository->get($this->getOrderId());
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Model/Order/Invoice.php:239]: Doc block type hint of `$_order` property was not updated accordingly.

File: app/code/Magento/Sales/Model/Order/Invoice.php [hunk 4]
```diff
@@ -96,9 +96,9 @@ class Invoice extends AbstractModel implements EntityInterface, InvoiceInterface
     protected $_invoiceConfig;
 
     /**
-     * @var \Magento\Sales\Model\OrderFactory
+     * @var \Magento\Sales\Api\OrderRepositoryInterface
      */
-    protected $_orderFactory;
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Model/Order/Invoice.php:101]: Removing `$_orderFactory` may break clients already relying on this property. It's better to deprecate it.

File: app/code/Magento/Sales/Model/Order/Invoice.php [hunk 5]
```diff
@@ -236,7 +236,7 @@ public function setOrder(\Magento\Sales\Model\Order $order)
     public function getOrder()
     {
         if (!$this->_order instanceof \Magento\Sales\Model\Order) {
-            $this->_order = $this->_orderFactory->create()->load($this->getOrderId());
+            $this->_order = $this->orderRepository->get($this->getOrderId());
         }
         return $this->_order->setHistoryEntityName($this->entityType);
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Model/Order/Invoice.php:241]: * Method doc block still states that return type is `Magento\Sales\Model\Order` which may not be true in some cases. It's best to change it to `OrderInterface` now. 
* Method clients must be checked to avoid unexpected consequences.
* `setOrder` method of this class is now inconsistent with `getOrder`. It's best to check possibility of aligning signatures and behaviors.

File: app/code/Magento/Sales/Test/Unit/Model/Order/InvoiceTest.php [hunk 6]
```diff
@@ -74,6 +75,11 @@ protected function setUp()
                 ]
             )
             ->getMock();
+        $this->orderRepository = $this->getMockBuilder(OrderRepository::class)
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Test/Unit/Model/Order/InvoiceTest.php:78]: It's best to mock `OrderRepositoryInterface` here

File: app/code/Magento/Sales/Test/Unit/Model/Order/InvoiceTest.php [hunk 7]
```diff
@@ -74,6 +75,11 @@ protected function setUp()
...
+            ->disableOriginalConstructor()
+            ->setMethods(['get'])
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Test/Unit/Model/Order/InvoiceTest.php:80]: What is the reason for `setMethods` here?

File: app/code/Magento/Sales/Test/Unit/Model/Order/InvoiceTest.php [hunk 8]
```diff
@@ -156,17 +160,13 @@ public function testGetOrderLoadedById()
     {
         $orderId = 100000041;
         $this->modelWithoutOrder->setOrderId($orderId);
-        $this->order->expects($this->once())
-            ->method('load')
-            ->with($orderId)
-            ->willReturnSelf();
+        $this->orderRepository->expects($this->once())
```
Comment (ishakhsuvarov) [inline app/code/Magento/Sales/Test/Unit/Model/Order/InvoiceTest.php:163]: It may be a good idea to check both cases: 
* in-object cache is used
* new entity is created

**Comment by ishakhsuvarov** [comment]
I would also suggest another testing scenario: verify that none of the currently existing properties in web api response is missing/changed after the change.

**Comment by sidolov** [comment]
@TomashKhamlai , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration!
